File: src/GameCore/Tag/TagRegistry.cs
using System.Text.Json;
using GameCore.Effect;

namespace GameCore.Tag;

/// <summary>
/// 标签注册表：从 JSON 加载全部 TagDef。
/// 设计契约层的数据容器，数据加载时校验 allowedSources，运行时不动。
/// </summary>
public sealed class TagRegistry
{
    /// <summary>全局单例引用，供 TagTierAtLeast 等条件查询。</summary>
    public static TagRegistry Instance { get; private set; }

    /// <summary>[适性]标签的涌现效果：所有层级条件要求 -N 级。</summary>
    public int TierReduction { get; set; }

    private readonly Dictionary<string, TagDef> _defsById = new();
    private readonly Dictionary<TagCategory, List<TagDef>> _defsByCategory = new();

    public void Load(string jsonPath, EffectParser effectParser)
        => LoadFromText(File.ReadAllText(jsonPath), effectParser);

    public void LoadFromText(string json, EffectParser effectParser)
    {
        using var doc = JsonDocument.Parse(json);
        foreach (var prop in doc.RootElement.EnumerateObject())
        {
            var id = prop.Name;
            var node = prop.Value;
            var def = new TagDef(
                id,
                node.TryGetProperty("name", out var n) ? n.GetString() : id,
                node.TryGetProperty("description", out var d) ? d.GetString() : "",
                ParseCategory(node),
                node.TryGetProperty("tier", out var t) ? t.GetInt32() : 1,
                ParseSources(node),
                effectParser.ParseEffects(node),
                ParseBehaviorWeights(node));
            Register(def);
        }
        Instance = this;
    }

    public void Register(TagDef def)
    {
        _defsById[def.Id] = def;
        if (!_defsByCategory.TryGetValue(def.Category, out var list))
            _defsByCategory[def.Category] = list = new List<TagDef>();
        list.Add(def);
    }

    public TagDef Get(string id)
    {
        if (!_defsById.TryGetValue(id, out var def))
            throw new ArgumentException($"未知标签: {id}");
        return def;
    }

    public bool TryGet(string id, out TagDef def) => _defsById.TryGetValue(id, out def);

    public IReadOnlyList<TagDef> GetByCategory(TagCategory category)
        => _defsByCategory.TryGetValue(category, out var list) ? list : Array.Empty<TagDef>();

    public IEnumerable<TagDef> GetAll() => _defsById.Values;

    private static TagCategory ParseCategory(JsonElement node)
    {
        var raw = node.TryGetProperty("category", out var c) ? c.GetString() : "ELEMENT";
        return raw.ToUpperInvariant() switch
        {
            "ELEMENT" => TagCategory.Element,
            "IDENTITY" => TagCategory.Identity,
            "PERSONALITY" => TagCategory.Personality,
            "EMOTION" => TagCategory.Emotion,
            "QUEST_MARK" => TagCategory.QuestMark,
            "SKILL" => TagCategory.Skill,
            _ => throw new ArgumentException($"未知标签分类: {raw}")
        };
    }

    private static IReadOnlySet<TagSource> ParseSources(JsonElement node)
    {
        var set = new HashSet<TagSource>();
        if (!node.TryGetProperty("allowedSources", out var arr)) return set;
        foreach (var s in arr.EnumerateArray())
        {
            set.Add(s.GetString().ToUpperInvariant() switch
            {
                "RACE" => TagSource.Race,
                "CLASS" => TagSource.Class,
                "QUEST" => TagSource.Quest,
                "TRAIT" => TagSource.Trait,
                "EQUIP" => TagSource.Equip,
                "BUFF" => TagSource.Buff,
                _ => throw new ArgumentException($"未知标签来源: {s.GetString()}")
            });
        }
        return set;
    }

    private static IReadOnlyDictionary<string, Dictionary<string, int>> ParseBehaviorWeights(JsonElement node)
    {
        var result = new Dictionary<string, Dictionary<string, int>>(StringComparer.OrdinalIgnoreCase);
        if (!node.TryGetProperty("behaviorWeights", out var bw)) return result;
        foreach (var pool in bw.EnumerateObject())
        {
            var dict = new Dictionary<string, int>();
            foreach (var opt in pool.Value.EnumerateArray())
                dict[opt.GetProperty("option").GetString()] = opt.GetProperty("weight").GetInt32();
            result[pool.Name] = dict;
        }
        return result;
    }
}

File: canglan-backend/canglan-core/src/main/java/com/canglan/core/tag/TagRegistry.java
package com.canglan.core.tag;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.canglan.core.effect.EffectParser;
import com.canglan.core.json.JsonReader;
import com.canglan.core.json.JsonValue;

/**
 * 标签注册表：从 JSON 加载全部 TagDef。对应 C# TagRegistry。
 * 数据加载时校验 allowedSources，运行时不动。
 */
public final class TagRegistry {

    /** [适性]标签的涌现效果：所有层级条件要求 -N 级。 */
    private int tierReduction;

    private final Map<String, TagDef> defsById = new LinkedHashMap<>();
    private final Map<TagCategory, List<TagDef>> defsByCategory = new HashMap<>();

    public void loadFromText(String json, EffectParser effectParser) {
        JsonValue root = JsonReader.parse(json);
        if (!root.isObject()) throw new IllegalArgumentException("tags.json 根节点必须是对象");
        for (Map.Entry<String, JsonValue> entry : root.asObject().entrySet()) {
            String id = entry.getKey();
            JsonValue node = entry.getValue();
            TagDef def = new TagDef(
                    id,
                    node.getString("name", id),
                    node.getString("description", ""),
                    parseCategory(node),
                    node.getInt("tier", 1),
                    parseSources(node),
                    effectParser.parseEffects(node),
                    parseBehaviorWeights(node));
            register(def);
        }
        // 注册为当前激活注册表（供 TagTierAtLeast 等条件访问）
        TagRegistryHolder.set(this);
    }

    public void register(TagDef def) {
        defsById.put(def.id(), def);
        defsByCategory.computeIfAbsent(def.category(), k -> new ArrayList<>()).add(def);
    }

    public TagDef get(String id) {
        TagDef def = defsById.get(id);
        if (def == null) throw new IllegalArgumentException("未知标签: " + id);
        return def;
    }

    /** 不存在返回 null（对应 C# TryGet）。 */
    public TagDef tryGet(String id) {
        return defsById.get(id);
    }

    public List<TagDef> getByCategory(TagCategory category) {
        return defsByCategory.getOrDefault(category, List.of());
    }

    public Collection<TagDef> getAll() {
        return defsById.values();
    }

    public int size() { return defsById.size(); }

    public int getTierReduction() { return tierReduction; }
    public void setTierReduction(int tierReduction) { this.tierReduction = tierReduction; }

    // ==================== 解析辅助 ====================

    private static TagCategory parseCategory(JsonValue node) {
        return TagCategory.parse(node.getString("category", "ELEMENT"));
    }

    private static Set<TagSource> parseSources(JsonValue node) {
        Set<TagSource> set = new HashSet<>();
        JsonValue arr = node.get("allowedSources");
        if (arr == null || !arr.isArray()) return set;
        for (JsonValue s : arr.asArray()) set.add(TagSource.parse(s.asString()));
        return set;
    }

    private static Map<String, Map<String, Integer>> parseBehaviorWeights(JsonValue node) {
        Map<String, Map<String, Integer>> result = new HashMap<>();
        JsonValue bw = node.get("behaviorWeights");
        if (bw == null || !bw.isObject()) return result;
        for (Map.Entry<String, JsonValue> pool : bw.asObject().entrySet()) {
            Map<String, Integer> dict = new HashMap<>();
            if (pool.getValue().isArray()) {
                for (JsonValue opt : pool.getValue().asArray()) {
                    dict.put(opt.getString("option", null), opt.getInt("weight", 0));
                }
            }
            result.put(pool.getKey(), dict);
        }
        return result;
    }
}
